OOP版lsコマンドの提出物

yocaji opened the pull request:
OOP版lsコマンドの提出物になります。

## 気になっている点

- `def initialize`の中でどこまで処理をやってしまっていいのか、書いているうちに分からなくなってきました
  - たとえば`Item`クラスでは、initialize以外のメソッドをすべてプライベートにして、initializeの中でそれらを呼び出して変数にセットして`attr_reader`でアクセスできるようにすればいいような気もしました

Commit:
OOP版lsコマンド作成

Commit:
テストで判明した点を修正した

Commit:
テスト一式を作成した

JunichiIto commented:
動画でレビューしたので別途リンクを送ります〜。

Commit:
レビューでいただいたコメントに基づいて修正した

ls.rb
  - Ls.newの実行処理をmainメソッドに移動した
  - ls.rbをrequire_relativeしたタイミングで実行されないようにした
  - 値の受け渡しのためにインスタンス変数を使わないようにした
  - Dir.chdirせずにパスとファイル名で処理できるようにした
item.rb
  - mtimeは加工せずに返すようにした
  - long_name(add_symlink)の処理をlong_format.rbに移動した
  - parse(format)_mtimeの処理をlong_format.rbに移動した
long_format.rb
  - アイテム数0の場合の条件分岐を廃止した
プログラム全体を通して、メソッド名・変数名・ブロック変数名を見直した

Commit:
プログラムの修正に合わせてspecを修正した

yocaji commented on 08.ls_object/long_format.rb:
```diff
@@ -5,29 +5,41 @@ def initialize(items)
+    end
+  end
+
+  def format_name(name, path)
```
シンボリックリンクの場合の処理を `item.rb` から `long_format.rb` に移動しました

yocaji commented on 08.ls_object/long_format.rb:
```diff
@@ -5,29 +5,41 @@ def initialize(items)
+    aligned_columns_without_last.concat([columns.last]).compact.transpose
+  end
+
+  def format_mtime(mtime)
```
最終更新日のフォーマット処理を `item.rb` から `long_format.rb` に移動しました

yocaji commented on 08.ls_object/ls.rb:
```diff
@@ -9,16 +10,14 @@ class Ls
-      LongFormat.new(@items).text
+      LongFormat.new(items).create_text
     else
-      return if @items.size.zero?
```
ここでの条件分岐をなくして、`short_format` 側で処理するようにしました

yocaji commented on 08.ls_object/ls.rb:
```diff
@@ -29,33 +28,37 @@ def fetch_items
 
-  def trim_names
-    @names.delete_if { |name| name.start_with?('.') }
+  def sort_all_items(target)
```
メソッド名を `bulk_names` から `sort_all_items` に変更しましたが、すべてのアイテム名を取得するという事が示せていないので、微妙な気がしています💦

yocaji commented on 08.ls_object/ls.rb:
```diff
@@ -29,33 +28,37 @@ def fetch_items
+  puts Ls.new(target, options).show
+end
+
+if __FILE__ == $PROGRAM_NAME
```
ご指摘の通り、`require_rerative`したタイミングでputsが実行されていたので修正しました。

yocaji commented on 08.ls_object/long_format.rb:
```diff
@@ -5,29 +5,41 @@ def initialize(items)
   end
 
-  def text
+  def create_text
```
メソッド名と変数名が同じにならないように修正しました

yocaji commented on 08.ls_object/long_format.rb:
```diff
@@ -5,29 +5,41 @@ def initialize(items)
+  def create_text
     text = ''
-    text += "total #{sum_blocks}\n"
-    if @items.size.positive?
```
アイテム数が0の時は`create_table`がnilを返すようにして、ここでの条件分岐をなくしました

yocaji commented on spec/08.ls_object/item_spec.rb:
```diff
@@ -48,18 +47,7 @@
-    example 'シンボリックリンクの時' do
-      slink = Item.new('slink')
-      expect(slink.long_name).to eq 'slink -> 123.txt'
+      expect(@item.mtime.to_i).to eq Time.local(2022, 5, 22, 14, 48, 47, 376666.732).to_i
```
Time型で比較したかったのですが以下のようにエラーとなってしまったため、ひとまず `to_i` をつけていますがもっと良い方法がありそうに思っています

```
Failures:

  1) Item#mtime 最終更新時刻を返す
     Failure/Error: expect(@item.mtime).to eq Time.local(2022, 5, 22, 14, 48, 47, 376666.732)

       expected: 2022-05-22 14:48:47.376666732 +0900
            got: 2022-05-22 14:48:47.376666732 +0900

       (compared using ==)

       Diff:
       @@ -1 +1 @@
       -2022-05-22 14:48:47 6471085181724787/17179869184000000 +0900
       +2022-05-22 14:48:47.376666732 +0900

     # ./spec/08.ls_object/item_spec.rb:50:in `block (3 levels) in <top (required)>'
```

yocaji reviewed (commented):
いただいたレビュー動画に沿って修正してみました。
何か認識違いしてそうな部分などありましたら、教えていただけると嬉しいです。
ご確認よろしくお願いいたします🙇‍♂️

Commit:
不要な処理を削除、rubocopの修正

JunichiIto reviewed (commented):
今回も動画でレビューしたので別途リンクを送ります〜。

Commit:
相対パスの指定をFile.expand_pathを使う方法に変更した

Commit:
mainメソッドをトップレベルからLsクラスのクラスメソッドに変更した

Commit:
Lsクラスの変数名とメソッド名を整理した

Commit:
ShortFormatクラスの変数名を修正、冗長な処理を修正

Commit:
align_table内のブロック変数名を修正した

Commit:
配列を活用して出力結果を作成するように修正した

- テストの期待値も修正した（以前の期待値の方が間違っていた）

Commit:
最終更新日時に関わるテストを修正

- 半年以内かどうかでの表示テストを追加した
- Item.mtimeをDate型のまま比較するようにした
- sampleディレクトリ内のファイル更新に伴って期待値が変わるテストを修正した

Commit:
create_tableの中の冗長な記述を削除した

Commit:
テストケース追加

- 名前なし引数がファイルの時のテストを追加した

yocajii commented:
下記2点を除いて、レビュー動画で教えていただいた点を修正してみました。
お手数ですが修正後のご確認と、以下の「未対応の項目について」と「その他補足」についてアドバイスいただけないでしょうか🙇‍♂️

**未対応**
1. Itemクラスの引数を`(name, path)`から`(path)`だけにする
1. long形式の時の整列は全体をまとめて処理にかけるか、必要な列だけ処理する

## 未対応の項目について

1については、`.`や`..`をpathにすると現状の書き方ではカレントディレクトリや親ディレクトリの名前になってしまい、そこから`.`や`..`を復元する方法の見当がつかず、別の方向としてそもそもpathとして渡すものを`～/..`のようにできないかという事も考えたのですがやはり見当がつかなかったという状況です。

2については、まず「全体をまとめて（最終列を外さずに）整列処理にかける」とすると日本語のファイル名がmaxとなる時に半角スペースの数が合わなくなったため外していました。
（が、これを書いているうちに見た目上は問題ないので、テストの期待値だけ変更すれば良いような気もしてしました）
そこで、必要な列だけを整列するというアプローチをとろうとしたのですが、個別に整列処理を書いて今よりも更に読みづらくなるような方向に迷走し始めてしまったので、ひとまず手を加えずに置いてあります。

## その他補足

ShortFormatクラスの`align_table`メソッドのブロック変数名が`|column|`ではなく`|columns|`ではないかという件ですが、ひとまず（まだマシそうな）`|row|`に変更しました。

以下の画像に描いたような考えでcolumnになる予定のデータということで`|column|`としていましたが、いくらなんでも無理があるなと思いました💦
![Untitled (Draft)-1](https://user-images.githubusercontent.com/33394676/171778145-d8f52c08-9030-4d0b-8e03-e50aa92d9424.jpg)

JunichiIto commented on 08.ls_object/ls.rb:
```diff
@@ -7,13 +7,24 @@
+    opt.on('-r') { |v| options[:r] = v }
+    target = opt.parse(ARGV)[0] || '.'
+
+    puts Ls.new(target, options).show
```
`self == Ls`なのでnewだけでよかったりします。

```suggestion
    puts new(target, options).show
```

yocaji replied:
修正しました！ 361a1bf

JunichiIto reviewed (approved):
1点コメントしていますが、それ以外のコメントは動画でお伝えします〜。

Commit:
self.main内でのnewの書き方を修正した

Commit:
pathだけを受け渡して必要な時だけファイル名を取り出すように修正した

Commit:
digit_size()をShortFormatクラスからItemクラスに移動した

Commit:
Item.nameをインスタンス変数からインスタンスメソッドに変更した

Commit:
Rubocopの指摘箇所を修正した

Commit:
ソート処理を修正した

JunichiIto commented on 08.ls_object/long_format.rb:
```diff
@@ -0,0 +1,45 @@
+      width = column_values.map { |value| value.to_s.size }.max
+      column_values.map { |value| value.is_a?(Integer) ? value.to_s.rjust(width) : value.ljust(width) }
+    end
+    aligned_columns_without_last.concat([columns.last]).compact.transpose
```
もしかすると以前も聞いたかもしれませんが、compactって書かないと何か問題があるんでしたっけ？

yocaji replied:
アイテム数が0の時（空のフォルダを指定した時など）に、`[nil]`となってしまって次の`transpose`ができなくなるのを何とかしたくてこうなってたようです💦

0アイテムの時は`create_table()`の呼び出す前の段階で、条件を明示した上で空の配列を返すように修正してみました。 e5b07cf

JunichiIto commented on 08.ls_object/item.rb:
```diff
@@ -0,0 +1,93 @@
+  private
+
+  def parse_ftype(ftype)
+    TYPES[ftype]
```
大した仕事をしていないのでメソッド化しなくてもいいかも？

yocaji replied:
修正しました🙏 353962f

JunichiIto reviewed (commented):
ほぼOKですが、ちょっとだけコメントさせてもらいましたー。

で、ソートの件ですが、ここはそこまでこだわる部分ではないので、「Debianではこういう動きになってた」ということを僕が理解できればそれでOKとします。（てか、こんな動きをするlsコマンドもあるのは知らなかった・・・）

ぶっちゃけ、凝ったソートはせずに「rオプションなしなら、Rubyのsortメソッドでソート、ありなら、sort.reverseでソート」でも十分です〜。（Mac版ならこれだけで要件が満たせますし）

Commit:
parse_ftype()の処理はメソッド化せずに使うよう修正した

Commit:
アイテム数0の時に空のテーブルを返す処理を行うタイミングを修正した

yocaji reviewed (commented):
ソートの件、承知いたしました。
丁寧に見てくださってありがとうございます🙏✨
コメントいただいた点を修正しましたので、ご確認いただけますでしょうか🙇‍♂️

JunichiIto commented on 08.ls_object/long_format.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def create_text
+    total = "total #{sum_blocks(@items)}"
+    table = @items.empty? ? [] : create_table(@items)
```
```suggestion
    table = create_table(@items)
```

のままにして、`create_table`が`[]`を返すようにした方がスッキリするな〜と思いました。

yocaji replied:
> `create_table`が`[]`を返す

早速ありがとうございます🙏 修正してみました！ 3621587
（意味を取り違えてたらすみません）

`create_table`が`[]`を返す方がスッキリする理由について、よく分かっていないので教えていただけますでしょうか🙇‍♂️
アイテム数がいくつかによって処理を分けるのはあくまで内部的な都合なので、パブリックメソッド側で気にするよりもプライベートメソッドで面倒を見た方が良いから、という考え方で合っていますか？

JunichiIto replied:
> という考え方で合っていますか？

うーん、僕の考えとは微妙に違うかも？
僕の考えはこんな感じでした。

- `create_table`に「テーブルを作ってね！」とお願いしたら、こっちは何も考えないで戻り値を受け取るようにしたい
- `@items.empty? ? [] :`は`@items`があれば条件分岐できるので、`create_table`側に持っていって判断するのも簡単そう
- 「`create_table`の引数には空の配列を渡してはならない」という制約があるより、「俺が何でも受け止めるからどーんと来い！」っていうメソッドの方が気軽に呼びやすい（呼び出し側が気を遣う、というのは避けたい）

[この記事](https://qiita.com/jnchito/items/49fe82364bca2f4f37f5)にも書いたのですが、「使う側に気を遣わせる」設計ではなく、 「作る側が気を遣う」設計の方が基本的に良いと思います。

yocaji replied:
詳しい説明をありがとうございます🙏✨
> 「使う側に気を遣わせる」設計ではなく、 「作る側が気を遣う」設計

という考え方がまだ徹底できていないなぁと分かりました😅
これからもコードを書くときに気にしてみたいと思います

また、ご提案いただいた早期リターンの書き方に修正しました 1067027

JunichiIto reviewed (approved):
1点コメントしましたが大きな問題ではないのでこれでOKです！🙆‍♂️

Commit:
アイテム数0の時はcreate_table()側で空の配列を返すように修正した

Commit:
アイテム数0の場合の結果を早期リターンを使って返すように修正した
